### PR TITLE
Added FORCE_COLOR and NO_COLOR environment variables

### PR DESCRIPTION
**Pull request by jhol**
This patch add support for the `FORCE_COLOR` and `NO_COLOR` environment variables which are a de-facto standard for overriding automatic behaviour in circumstances where colors are required in redirected (non-TTY) output streams, or if colors are required to be removed in non-redirected (TTY) output streams.

### CODE REVIEW

**Comment by jhol** [comment]
By coincidence, this merge request also feeds into the wider `NO_COLOR` effort going on at https://no-color.org/

**Comment by 1138-4EB** [comment]
Hi @jhol! Thanks for proposing this PR and for the reference to no-color.org. I'm coming from tox-dev/tox#1468, and I wonder if you'd consider checking PY_COLORS too. Should you consider that `PY_COLORS` is not general enough and that the issue should be handled elsewhere, I'd be ok with it. I'm just looking for other thoughts on the point.

**Comment by jhol** [comment]
@1138-4EB You're welcome to start a branch on top of mine and add an additional patch to add the functionality you're looking for. Should be pretty simple - you just need to modify the line

```
no_color = 'NO_COLOR' in os.environ.keys()
```

It should be simple. I'd do it myself, but I don't have time to do the testing right now.

**Comment by jhol** [comment]
@tartley is this PR likely to be accepted?

**Comment by jhol** [comment]
@tartley - sorry to nag. Do you know if you will have time to look at this any time soon?

**Comment by tartley** [comment]
Hi, @jhol Thanks heaps for the PR. Apologies I hadn't noticed it previously. It sounds eminently reasonable. I'll take a look at it...

**Comment by eine** [comment]
> I was reminded recently that many people rely on Colorama, when I accidentally broke a release and was instantly deluged with issue reports, emails and tweets.

That sound really humble, considering that there are [~72k dependents](https://github.com/tartley/colorama/network/dependents?package_id=UGFja2FnZS01MDk2OTU1OQ%3D%3D).

> So we'll have to think about this carefully. Either you can do it, and write about it to demonstrate that you definitely understand what's going on, or else I can have a think about it sometime and try to figure it out.

I can contribute a GitHub Actions (GHA) workflow that would test this in:

- Windows:
  - Cmd + Python native
  - Powershell + Python native
  - Git for Windows + Python native
  - MSYS2 + Python native
  - Cmd + Python from MSYS2
  - Powershell + Python from MSYS2
  - Git for Windows + Python from MSYS2
  - MSYS2 + Python from MSYS2
- GNU/Linux
- macOS

Unfortunately, GHA does neither provide nor fake a TTY (actions/runner#241). Hence, the GNU/Linux environment is the only one where we can get one (in a container). There are two options to test the effect of `FORCE_COLOR`, `NO_COLOR` and `<>` in this context:

- Add some mechanism to the codebase to override `isatty` for testing purposes.
- https://github.com/actions/runner/issues/241#issuecomment-577370913

What do you think?

File: colorama/ansitowin32.py
Comment (jhol) [reply colorama/ansitowin32.py]: Are you sure about this?

In the comment lower down it says "Write the given text to our wrapped stream, stripping any ANSI sequences from the text, and optionally converting them into win32 calls." - so if `NO_COLOR` is specified, for example, we would want colorama to strip the color sequences off any incoming strings.

Or do I misunderstand the meaning?

Comment (tartley) [reply colorama/ansitowin32.py]: Ah, sorry, I think the name "strip" has been confusing.

If I remember correctly, the name means this:

In normal operation, on Windows, we both "convert" ANSI characters to win32 API calls (thus displaying colors in the terminal), AND then "strip" the ANSI chars from the output (to prevent the ANSI characters from being echoed to the output in the terminal)

So whether a user wants to see colors or not, on Windows we always want to strip the ANSI from the output.

**Comment by tartley** [comment]
@eine That does sound fabulous, thank you for the thoughts.

The 'actions/runner' option sounds promising, but seems to be stalled on exactly how to implement it? Or am I misreading?

The 2nd option could at least partially lean on *unit* tests to exercise the details of this code even if no tty is actually present. (My memory of this code is vague, but I would hope that is what is currently present)

eg1. Pass fake streams that will look like they do or do not have a tty:

```
stream_with_tty = Mock(closed=False, isatty=lambda: True)

def test_xxx(self):
    ansiToWin32 = AnsiToWin32(stream_with_tty)
```

eg2. Or maybe inject a fake StreamWrapper class?

```
fake_sw = Mock(closed=False, isatty=lambda _: True)

@patch("colorama.ansitowin32.StreamWrapper", fake_sw)
def test_xxx(self)
    # inside this constructor call, "isatty" will evaluate to True
    ansiToWin32 = AnsiToWin32(...)
```

But, you are right, it would be nice to ALSO have some end to end test that sanity checks a couple of core scenarios with and without a tty present. I think when I first wrote this, I punted on that, and wrote "demo" scripts instead, which should produce relatively easy-to-check output for humans. 🤮

File: colorama/ansitowin32.py
Comment (tartley) [reply colorama/ansitowin32.py]: Regarding step (3), I might be mistaken than we can eliminate "self.stream.closed" (I just don't know.) Maybe it's a necessary guard for the terms that come afterwards? So if we have to keep this long line, we can just format it, and that's fine. Something like:

```
convert = (
    conversion_supported and
    not self.stream.closed and
    (isatty or force_color) and
    not no_color
```

Comment (jhol) [reply colorama/ansitowin32.py]: Ok I've made the changes you suggested... I think I managed to include everything. What do you think?

**Comment by eine** [comment]
@tartley, as a starting point, I created #249. It does not run any "mocked" check yet. Instead, the demos are executed. The next step is to write a proper test.

However, I based the PR on this one, so I added three tests to each job. I found that `FORCE_COLOR` and `NO_COLOR` seem to be ignored. See, for example, https://github.com/eine/colorama/runs/418587017 @jhol, any guess?

**Comment by jhol** [comment]
@eine I reset the PR back to my original submission without @tartley 's suggestions, because they prevent the patches from working correctly. I don't know how we can resolve this, because as far as I can see... everything in my patch-set is necessary for it to work as intended:

Here is a series of tests:

- Running in a Linux terminal
- Running in a Linux terminal with `NO_COLOR` set
- Piping to a file on Linux
- Piping to a file on Linux with `FORCE_COLOR` set
- Running in a windows console window
- Running in a windows console window with `NO_COLOR` set.

20200212-colorama-no-color-test

In addition, I have tested:

- Running in a Msys2 terminal
- Running in a Msys2 terminal with `NO_COLOR` set
- Piping to a file on Msys2
- Piping to a file on Msys2 with `FORCE_COLOR` set

...and in each case the result is correct.

How can we resolve the gap between the comments received on this PR and the tests I have done that pass?

**Comment by eine** [comment]
@jhol I rebased my PR on top of your branch, and on top of master. I can confirm that it now works as expected for all the tested cases: https://github.com/eine/colorama/actions/runs/38098545 

> How can we resolve the gap between the comments received on this PR and the tests I have done that pass?

Ideally:

1. Rebase this PR on top of master.
2. That programatic checks to #249 and merge it.
3. Rebase this PR on top of #249 and add corresponding programatic checks.
4. Apply @tartley's proposals one by one. Some of them will break the checks.
5. When all the tests are green, merge this PR.

Alternatively, we can skip adding programatic checks to #249, and/or merging it. You can still rebase on top of it to have all the tests executed automatically when you push. However, you need to visually check all the jobs.

**Comment by jhol** [comment]
@eine 

> 1. Rebase this PR on top of master.

Done. Specifically rebased on top of your #249 patch.

> 2. Add programatic checks to #249 and merge it.

See above.

> 3. Rebase this PR on top of #249 and add corresponding programatic checks.

Would you be able to handle this? It's a bit beyond the scope of what I have time to get into right now.

> 4. Apply @tartley's proposals one by one. Some of them will break the checks.

The logic certainly was certainly getting rather confusing, so I used Yosys to optimize the logic:

20200212-colorama-optimized-logic

It's a little better than it was before, but it's still a rather convoluted set of decisions - irreducibly so.

Despite what @tartley said, it won't work if I take out the changes to the `strip` decision.

> 5. When all the tests are green, merge this PR.

The results of the [test runs](https://github.com/vcatechnology/colorama/runs/441675773) and the test suite are both passing now.

**Comment by tartley** [comment]
Acknowledged. thanks for all the work on this. Sorry it turned out to be a pain. It's late here, and I don't trust myself to give it the final review it needs. I'll look at it tomorrow. Thank you!

**Comment by eine** [comment]
> > 1. Rebase this PR on top of #249 and add corresponding programatic checks.
> 
> Would you be able to handle this? It's a bit beyond the scope of what I have time to get into right now.

Unfortunately, I'm not familiar with either writing tests with `Mock` or colorama's codebase; and I cannot allocate time to learn it ATM. I can help with modifying the GHA workflow files to replace `demo.sh` and/or to set up finer grained execution/reports. Nonetheless, I need someone else to step ahead and "modernise" the test scripts.

> The logic certainly was certainly getting rather confusing, so I used Yosys to optimize the logic:

This is quite clever! May I know how did you do it? Did you write an HDL version of the logic?

> The results of the [test runs](https://github.com/vcatechnology/colorama/runs/441675773) and the test suite are both passing now.

I think that this PR can be merged as is, and testing scripts can be improved later.

**Comment by jhol** [comment]
@eine 

> This is quite clever! May I know how did you do it? Did you write an HDL version of the logic?

Glad you asked.

I wrote up the old logic in a verilog file:
```verilog
module colorama(
  input wire conversion_supported,
  input wire closed,
  input wire stream_isatty,
  input wire force_color,
  input wire no_color,
  output wire strip,
  output wire convert);

  wire isatty = !closed && stream_isatty;
  wire color_allowed = (isatty || force_color) && !no_color;

  assign strip = conversion_supported || (!closed && !color_allowed);
  assign convert = conversion_supported && (!closed && color_allowed);
endmodule
```

Then ran yoysys to load the design, optimize it and show a graphviz dot graph of the result:
```
yosys -p "read_verilog colorama.v; opt; show colorama"
```

I feel like this was complete overkill for a boolean-algebra simplification that I should have easily done with pen and paper - but hey, it's not stupid if it works.

**Comment by jhol** [comment]
Bump?

**Comment by eine** [comment]
@jhol, I'm sorry I forgot to reply. Thanks a lot for the explanation. Being on the VHDL side, I think I would have not thought about such a solution because the syntax is quite different. Nonetheless, I believe it's not stupid. Actually, you kind of did it with pen and paper already, and this was a checker :D

**Comment by tartley** [comment]
I am on vacation and unlikely to look at this until next week. Hugs!

**Comment by jhol** [comment]
Hi! How was the vacation?

**Comment by tartley** [comment]
Hey. It was good, but I'm in Europe for work this week. I bet I can find some time to look though. Thanks for the ping!

**Comment by tartley** [comment]
Also, I don't mean to be a lone gatekeeper here. If @wiggin15 has an opinion, please feel free to review and merge. Otherwise I'll get to it soon... Thank y'all!

File: README.rst
Comment (wiggin15) [inline README.rst]: IMHO there are a little too many "and"s and "or"s in this sentence and it would be good to break it down for clarity.

File: colorama/ansitowin32.py
Comment (wiggin15) [inline colorama/ansitowin32.py]: I may be wrong, but does `not_closed` have meaning here? We check for `not_closed` later, for `strip` and `convert` again, and I think that the expressions there mean we can skip the check for `not_closed` here.
(if we find out we only need to check it in one place - I prefer to leave the full expression, `not self.stream.closed`, because negative booleans, i.e. any `not_*` variables, are kinda confusing IMO)

Comment (jhol) [reply colorama/ansitowin32.py]: Yes - it is rather confusing, but in terms of the logical structure, it appears to be optimal, in that I optimized it using Yosys (see above).

File: README.rst
Comment (jhol) [reply README.rst]: I could rewrite the comment as a series of bullet-points

File: colorama/ansitowin32.py
Comment (wiggin15) [reply colorama/ansitowin32.py]: I'm not familiar with Yosys, and it appears that it didn't change the original logic. From my "pen and paper" calculations (and also a test with a small Python script) I still think that removing `not closed` from the `color_allowed` expression should not affect the result or `convert` and `strip` regardless of the values of the inputs. Maybe Yosys only optimizes logic if there are more complex changes like removing inputs or something like that? Or I missed something?

Comment (jhol) [reply colorama/ansitowin32.py]: Ok... I've just pushed a branch with the following logic:

```
        not_closed = not self.stream.closed
        force_color = 'FORCE_COLOR' in os.environ.keys()
        no_color = 'NO_COLOR' in os.environ.keys()
        color_allowed = (self.stream.isatty() or force_color) and not no_color
```

but it results in [test errors](https://travis-ci.org/tartley/colorama/builds/657729965?utm_source=github_status&utm_medium=notification), because you can't do `isatty` on a closed stream! So I think the logic has to stay as-is - though I think everyone who's looked at this code found it confusing.

So I decided to break it up with more comments.

File: README.rst
Comment (jhol) [reply README.rst]: Ok - I rewrote it as a set of bullet-points. It seems slightly easier to read now.

**Comment by jhol** [comment]
I think I've addressed all the points of criticism as best as possible, and all the tests pass:

- https://github.com/vcatechnology/colorama/actions/runs/48800182
- https://travis-ci.org/tartley/colorama/builds/657733296

Can this merge now?

**Comment by jhol** [comment]
Bump?

File: colorama/ansitowin32.py
Comment (eric-wieser) [inline colorama/ansitowin32.py]: Possibly clearer as:
```
if  'NO_COLOR' in os.environ:
    color_allowed = False
elif 'FORCE_COLOR' in os.environ:  # overriden by NO_COLOR
    color_allowed = True
else:
    color_allowed = isatty
```

Comment (jhol) [reply colorama/ansitowin32.py]: Is this clearer? Just seems more verbose to me. But if changing it would get this PR over the line, I'm happy to make the change.

Comment (eric-wieser) [reply colorama/ansitowin32.py]: It makes it clear that there's a priority order of things to look at. Clearness is certainly not measured by character count alone ;).

I'm not part of this project though, so my review only matters if a maintainer says it does. I'm here because I received a PR for a project I maintain that added `NO_COLOR`, and wanted to see how it should interact with `FORCE_COLOR`.

Comment (jhol) [reply colorama/ansitowin32.py]: Hey thanks for taking the time to review the PR.

And ok... you've talked me into it. I made the change. Your `'FORCE_COLOR' in os.environ` is superior to my 'FORCE_COLOR' in os.environ.keys()`.

Hopefully we can get this merged as soon as possible. Right now this PR feels like it's stuck in bike-shed hell.

**Comment by tartley** [comment]
I hear you. Apologies, that's at least partly my fault. I'll find time today to progress it, hopefully to merge. Thank you for all your work, it really is very much appreciated.

File: colorama/ansitowin32.py
Comment (tartley) [reply colorama/ansitowin32.py]: I hear you. Apologies, that's at least partly my fault. I'll find time today to progress it, hopefully to merge. Thank you for all your work, it really is very much appreciated.

Comment (jhol) [reply colorama/ansitowin32.py]: No worries I quite understand - I'm just keen to get this closed out, and I think it would save all our time if we could reduce the number of round trips, so if there's anything you want me to do, just lay it out - or even feel free to add your own patches, or amend mine.

It sounds like the concept of this PR is acceptable. In which case it's just a question of whether the implementation is sufficient in quality. The code has gone through a few iterations now, so it seems pretty good to me, but let me know if there are any desired fixes.

**Comment by jhol** [comment]
:mask: :mask: :mask: Anyone around to approve this? :mask: :mask: :mask:

@tartley  @eine  ?

**Comment by tartley** [comment]
Hi there. I know we'd love to merge it. But it's going to require careful checking. The last trivial update I pushed out (a README update, I think?) broke packaging and hence broke "pip install" for zillions of people. So it will need careful checking, and possibly an overhaul of our release process (eg. formal checklist for releasing to test.pypi first)

I personally am a bit snowed under with real life right now, and I imagine others are too. So I don't know when it'll happen.

If you're keen to use a release with this branch in it, then consider releasing it to PyPI yourself for the moment, under a different name, eg "colorama-force". There are probably targets in the Makefile that can help with that.

**Comment by jhol** [comment]
> Hi there. I know we'd love to merge it. But it's going to require careful checking. The last trivial update I pushed out (a README update, I think?) broke packaging and hence broke "pip install" for zillions of people. 

I'm very sympathetic - maintaining a component with so many deps is a huge responsibility!

> So it will need careful checking, and possibly an overhaul of our release process (eg. formal checklist for releasing to test.pypi first)

I included @eine 's [ci: add GHA workflow 'test' (`c5dfed525`)](https://github.com/tartley/colorama/pull/230/commits/c5dfed525066f679ae9e9497a755321e5b378758) commit in my pull request. You can include it or not as you choose, but when I used it to run tests, it validated my commit successfully.

I'm slightly fuzzy on what it would take to properly tie this into the into the project infrastructure. I suppose that's a question for @eine.

As for my commit, it's pretty trivial now: https://github.com/tartley/colorama/pull/230/commits/06292218f9ae475f6da32b33d5f041716a37b2e1

With @eric-wieser 's comments applied, it's quite straightforward now - there's no more confusing boolean mathematics.

> I personally am a bit snowed under with real life right now, and I imagine others are too. So I don't know when it'll happen.

Yup - crazy times. I hope you're keeping well.

> If you're keen to use a release with this branch in it

It's being used by a build-server, so we have been installing it with `pip install` and the URL of my GitHub repository. For our needs it works fine, but I'm keen for this commit not to be lost in limbo - particularly because I think the [no-color project](https://no-color.org/) would life to have it upstream.

**Comment by tartley** [comment]
Acknowledged, and thank you for your understanding.

I too am very keen for this not to get lost. I *will* make sure it happens, just not today.

Best.

**Comment by eine** [comment]
@jhol, my knowledge is related to CI and precisely the details/annoyances/workarounds for GitHub Actions. Unfortunately, I'm not confident at all with Python and I cannot devote now learning how to do the testing/mocking properly (i.e. to make it fail programatically).

Nonetheless, I'd like to drop my cent here: would it be feasible for colorama to adopt a release-candidate publishing scheme? By making use of semantic versioning, it should be feasible to release features such as this one, without breaking workflows downstream (as long as consumers of this project use semver properly). The benefit of such an approach would be that part of the user base could have these features tested (and bugs reported) before actually making the change available to all the consumers.

File: colorama/ansitowin32.py
Comment (eric-wieser) [inline colorama/ansitowin32.py:104]:
```suggestion
            strip = conversion_supported or not color_allowed
```

File: colorama/ansitowin32.py
Comment (eric-wieser) [inline colorama/ansitowin32.py:109]:
```suggestion
            convert = conversion_supported and color_allowed
```

File: colorama/ansitowin32.py
Comment (eric-wieser) [inline colorama/ansitowin32.py:100]:
```suggestion
        # check if the FORCE_COLOR environment variable has been set and NO_COLOR has not been set.
        if 'NO_COLOR' in os.environ:
            color_allowed = False
        elif 'FORCE_COLOR' in os.environ:  # overriden by NO_COLOR
            color_allowed = True
        else:
            # can't ask if it's a tty if it's closed
            color_allowed = not self.stream.closed and self.stream.isatty()
```

Comment (jhol) [reply colorama/ansitowin32.py]: Makes sense. No need to put `isatty` in a variable

**Comment by tartley** [comment]
Hey. FYI, Yesterday I created [a PR to test releases](https://github.com/tartley/colorama/pull/276) before we push them to PyPI. When that is merged, I'll be more confident about resuming merges and releases. I'll try to look at this PR soon. Thank you for creating it!

**Comment by jhol** [comment]
:+1:

**Comment by jhol** [comment]
@tartley Did you make any progress in resurrecting your release process?

**Comment by tartley** [comment]
@jhol A release process was drafted, in README-hacking, and, if memory serves, we used it once, for the last release, months ago.  Other than that, I haven't made any time to look at Colorama. Let me confer with the other maintainer, @wiggin15

**Comment by jhol** [comment]
Hi! Is there any chance of getting this merged and released?

**Comment by tartley** [comment]
OK, this is brilliant, and has the added bonus that it seems to work! :-) Please let me know if I've misunderstood the following:

* This branch is, in your opinion, good to merge as-is.
* It's based of quite an old commit (I know, I know) and has a minor conflict in README.rst, which looks straightforward to merge.
* This branch now includes everything that was in https://github.com/tartley/colorama/pull/249, so we can delete #249 after merging this.

In order to get this done and not kick off a new round of delaying discussion, I might merge this, and then make some tiny nitpick changes myself on top of it:

1. In the README, use parens or something to be precise about the precedence of the 'and' and 'or' in the new description of when conversion/stripping happen.
2. Rename or very minor reformulation of the variable 'not_closed', to conform with Code Complete, page 269, *Use positive boolean variable names*.

@jhol Your thoughts appreciated, and if this sounds good let's merge this tomorrow, while I scan what other PRs ought to be included in any imminent release.

**Comment by jhol** [comment]
@eine Thanks for resolving the merge conflicts. I fast-forwarded this branch to include your merge-commit.

**Comment by jhol** [comment]
@eine @tartley - Ok this passes the build server now except for one of the macos builds: https://github.com/vcatechnology/colorama/runs/4225500386?check_suite_focus=true

**Comment by hugovk** [comment]
> @eine @tartley - Ok this passes the build server now except for one of the macos builds: [vcatechnology/colorama/runs/4225500386?check_suite_focus=true](https://github.com/vcatechnology/colorama/runs/4225500386?check_suite_focus=true)

Need to do something like https://github.com/tartley/colorama/pull/330 to fix PyPy3 on macOS.

**Comment by jhol** [comment]
> > @eine @tartley - Ok this passes the build server now except for one of the macos builds: [vcatechnology/colorama/runs/4225500386?check_suite_focus=true](https://github.com/vcatechnology/colorama/runs/4225500386?check_suite_focus=true)
> 
> Need to do something like #330 to fix PyPy3 on macOS.

Yes - that looks like it would fix it. Hopefully we can get this merged soon.

**Comment by EthanHindmarsh** [comment]
Any updates on this?

**Review by ssbarnea — APPROVED** [review]

**Comment by ssbarnea** [comment]
With all due respect for colorama, but why do we need to commemorate getting this simple and logic PR into its 3rd year of its life?

This being said, I am not surprised rich newer versions [removed colorama](https://github.com/Textualize/rich/issues/1991).

**Comment by tartley** [comment]
@ssbarnea Also, while large amounts of the blame rest on me for not being responsive to PRs like this, also, this particular PR has had real problems through it's life. I've been very explicit in my comments above, that it needs automated tests, and it still has none, so should not be merged. If it had tests, I like to think I would have been merged by now.

Firing up 'demo.sh' for humans to eyeball the results is not an appropriate replacement.

Merging this as-is, as with any untested code, simply isn't fair on future developers. It is punting the work that should be done in this MP onto other people. Years from now, someone else will be trying to merge some innocuous looking changes that break this code. This is inevitable. Nobody can write code perfectly that never breaks anything. But we are saved from committing our mistakes, because the tests warn us. Without tests to warn us that this code has been broken, other developers will have no guard rails to prevent them from breaking the code in this MP. Then, when a broken release goes out, it's me, or some other innocent third party, who will be dealing with the tsunami of issues, emails an tweets from angry and desperate users.

**Comment by ssbarnea** [comment]
@tartley Thanks for these awesome words. Yep, I know that maintaining is 99% pain, I have a long list of things that mostly depend on me... and I know that there is a category of entitled users ready to troll the maintainer about things he did wrong or did not do. Those do a lot of damage to us.

While I do have a lot of experience with packaging and testing (GHA) on python packages, I am not sure if I should raise my hand for colorama, even if I used it for years, mostly because the amount of time I spend on Windows is likely 2%, the rest being Windows and MacOS.

What scares me the most is that I do not see any checks reporting on this pull-request... without active CI I would be really afraid to do anything.

**Comment by ssbarnea** [comment]
I still do not see GHA running on this PR and I only original OP or a maintainer can use the close/open trick to force the pipelines to run again.

There is also another aspect: colorama is one of the very few projects that did not remove support for obsolete versions of python. This prevents the project from moving forwards as most of the tools/libraries around dropped support for it. Most attempts to fix something will get stuck on problems like this. I would not be surprised to see python 2.7 removed from GHA soon too, so it will be impossible to even test it.

Someone needs to be able to make decisions in order to keep this project alive, or it will be killed by strangulation. Bump major version and drop anything that is outdated, allow others to contribute and give the project a new life. Keeping a `stable/..` branch would be enough in order to allow potential emergency hotfixes for ancient versions.

I personally used to keep compatibility for very long time but in 2022 I managed to move most of the projects I maintain to require python 3.8+, and some of them have more github stars than colorama. It is bit fun as at beginning of 2021, I was still having projects supporting py27 on the main branch,... so dropped a big number of python version in a short period of time. I mentioned this because the boost in project ease of maintain was immense. The days of hoarding EOL python versions are gone. If someone really wants to backport a fix, they have to do the leg-work. Based on my experience, almost nobody did the work, they we motivated to spent the time upgrading their systems instead.

As long project uses semantic versioning, people understand what a major version mean. Those that installed dependencies without any kind of version ranges are supposed to read the results of their own mistakes.

**Comment by jhol** [comment]
> Also, while large amounts of the blame rest on me for not being responsive to PRs like this, also, this particular PR has had real problems through it's life. I've been very explicit in my comments above, that it needs automated tests, and it still has none, so should not be merged. If it had tests, I like to think it would have been merged by now.

I'm happy to write a test - I just don't really know how to go about it.

**Comment by kasperk81** [comment]
> https://no-color.org/

but FORCE_COLOR has no such effort. the closest thing to documentation is
https://nodejs.org/api/cli.html#force_color1-2-3
and it is not just "setting this variable to _any_ value" as NO_COLOR spec describes, rather its values are more involved: 1 => 16 color,  2 => 256 color, 3 => 16 million colors. any other value _disables_ the color if it was enabled (which is unnecessary when in the same documentation they are also supporting NO_COLOR ...). node.js is a good example of how to over engineer stuff.

**Review by tartley — CHANGES_REQUESTED** [review]
Needs automated tests

**Comment by tartley** [comment]
I don't think we can merge this. It mingles different kinds of changes and includes no tests.  The investigation and thought has been valuable, but it has also exposed that this change is complicated. Because of that, I'm going to close this PR as too risky.  If someone wants to resurrect the idea in a new MP, with tests, without blending in other changes like the GHA, then I'd be receptive. Let's talk about it before you do the work though.

**Comment by jeremy-rifkin** [comment]
Libraries like https://github.com/chalk and https://github.com/Textualize/rich support these, in case it helps

**Pull request closed without merging**
